addons/account_asset/account_asset_invoice.py

```diff
@@ -64,13 +64,29 @@ def asset_create(self):
                 asset.validate()
         return True
 
+    @api.onchange('asset_category_id')
+    def onchange_asset_category_id(self):
+        if not self.asset_category_id:
+            self.account_id = self.get_invoice_line_account(self.invoice_id.type, self.product_id, self.invoice_id.fiscal_position_id, self.invoice_id.company_id)
+        if self.invoice_id.type == 'out_invoice' and self.asset_category_id:
+            self.account_id = self.asset_category_id.account_asset_id.id
+        elif self.invoice_id.type == 'in_invoice' and self.asset_category_id:
+            self.account_id = self.asset_category_id.account_depreciation_id.id
+
+    @api.onchange('uom_id')
+    def _onchange_uom_id(self):
+        result = super(AccountInvoiceLine, self)._onchange_uom_id()
+        self.onchange_asset_category_id()
+        return result
+
     @api.onchange('product_id')
-    def onchange_product_id(self):
-        if self.product_id:
-            if self.invoice_id.type == 'out_invoice':
-                self.asset_category_id = self.product_id.product_tmpl_id.deferred_revenue_category_id
-            elif self.invoice_id.type == 'in_invoice':
-                self.asset_category_id = self.product_id.product_tmpl_id.asset_category_id
+    def _onchange_product_id(self):
```

Comment:
setting of the asset category on invoice from sales and purchase is to be done from the sales and purchase module.